Encode generated path parameters

Parameter generation based on patterns in path parameters can take values containing slashes (or maybe other characters that would change the meaning of the URI) that are not URL encoded.
That leads to false errors when testing the backend.

Example:
The following specification:
```yaml
openapi: "3.0.0"
paths:
    /foo/{param}/bar:
        parameters:
            - schema:
                type: string
                pattern: ......
              name: param
              in: path
              required: true
        post:
            summary: post stuff
            responses:
                200:
                    description: OK
```

The `status_code_conformance` test fails as follows:

```
(venv) (user@machine) schemathesis » schemathesis run /tmp/openapi.yml --base-url http://localhost:4010 -c status_code_conformance
==================================== Schemathesis test session starts ====================================
platform Linux -- Python 3.7.3, schemathesis-0.14.0, hypothesis-4.44.2, hypothesis_jsonschema-0.9.9, jsonschema-3.1.1
rootdir: /home/yoann/Downloads/schemathesis
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/yoann/Downloads/schemathesis/.hypothesis/examples')
Schema location: /tmp/openapi.yml
Base URL: http://localhost:4010
Specification version: Open API 3.0.0
collected endpoints: 1

POST /foo/{param}/bar F                                                                            [100%]

================================================ FAILURES ================================================
_________________________________________ POST: /foo/{param}/bar _________________________________________
Check           : status_code_conformance
Path parameters : {'param': '0000/0'}
Body            : {}

Run this Python code to reproduce this failure: 

requests.post('http://localhost:4010/foo/0000/0/bar', json={})
================================================ SUMMARY =================================================

status_code_conformance            4 / 14 passed           FAILED 

=========================================== 1 failed in 0.22s ============================================
```

because the generated value for `param` contains a `/`.

On server side we can clearly see that it was never url-encoded:

```
    [HTTP SERVER] post /foo/0000/0/bar ✖  error     Request terminated with error: https://stoplight.io/prism/errors#NO_PATH_MATCHED_ERROR: Route not resolved, no path matched
```

I will try to have a look myself, but I'm completely unfamiliar with the codebase, do I don't promise anything.

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -16,6 +16,11 @@ Added
 - Pre-run hooks for CLI. `#147`_
 - A way to register custom checks for CLI via ``schemathesis.register_check``. `#270`_
 
+Fixed
+~~~~~
+
+- Not encoded path parameters. `#272`_
+
 Changed
 ~~~~~~~
 
@@ -393,6 +398,7 @@ Fixed
 .. _0.3.0: https://github.com/kiwicom/schemathesis/compare/v0.2.0...v0.3.0
 .. _0.2.0: https://github.com/kiwicom/schemathesis/compare/v0.1.0...v0.2.0
 
+.. _#272: https://github.com/kiwicom/schemathesis/issues/272
 .. _#270: https://github.com/kiwicom/schemathesis/issues/270
 .. _#268: https://github.com/kiwicom/schemathesis/issues/268
 .. _#261: https://github.com/kiwicom/schemathesis/issues/261

--- a/src/schemathesis/_hypothesis.py
+++ b/src/schemathesis/_hypothesis.py
@@ -4,6 +4,7 @@ import re
 from base64 import b64encode
 from functools import partial
 from typing import Any, Callable, Dict, Generator, Optional
+from urllib.parse import quote_plus
 
 import hypothesis
 import hypothesis.strategies as st
@@ -112,7 +113,7 @@ def get_case_strategy(endpoint: Endpoint) -> st.SearchStrategy:
     static_kwargs = {"path": endpoint.path, "method": endpoint.method, "base_url": endpoint.base_url}
     try:
         strategies = {
-            "path_parameters": from_schema(endpoint.path_parameters),
+            "path_parameters": from_schema(endpoint.path_parameters).filter(filter_path_parameters).map(quote_all),
             "headers": from_schema(endpoint.headers).filter(is_valid_header),  # type: ignore
             "cookies": from_schema(endpoint.cookies),
             "query": from_schema(endpoint.query),
@@ -121,6 +122,19 @@ def get_case_strategy(endpoint: Endpoint) -> st.SearchStrategy:
         return _get_case_strategy(endpoint, static_kwargs, strategies)
     except AssertionError:
         raise InvalidSchema("Invalid schema for this endpoint")
+
+
+def filter_path_parameters(parameters: Dict[str, Any]) -> bool:
+    """Single "." chars are excluded from path by urllib3.
+
+    In this case one variable in the path template will be empty, which will lead to 404 in most of the cases.
+    Because of it this case doesn't bring much value and might lead to false positives results of Schemathesis runs.
+    """
+    return not any(value == "." for value in parameters.values())
+
+
+def quote_all(parameters: Dict[str, Any]) -> Dict[str, Any]:
+    return {key: quote_plus(value) if isinstance(value, str) else value for key, value in parameters.items()}
 
 
 def _get_case_strategy(

--- a/test/app/__init__.py
+++ b/test/app/__init__.py
@@ -23,6 +23,7 @@ class Endpoint(Enum):
     success = ("GET", "/api/success", handlers.success)
     failure = ("GET", "/api/failure", handlers.failure)
     slow = ("GET", "/api/slow", handlers.slow)
+    path_variable = ("GET", "/api/path_variable/{key}", handlers.success)
     unsatisfiable = ("POST", "/api/unsatisfiable", handlers.unsatisfiable)
     invalid = ("POST", "/api/invalid", handlers.unsatisfiable)
     flaky = ("GET", "/api/flaky", handlers.flaky)
@@ -93,7 +94,8 @@ def make_schema(endpoints: Tuple[str, ...]) -> Dict:
         "paths": {},
     }
     for endpoint in endpoints:
-        method = Endpoint[endpoint].value[0].lower()
+        method, path, _ = Endpoint[endpoint].value
+        path = path.replace(template["basePath"], "")
         if endpoint == "unsatisfiable":
             schema = {
                 "parameters": [
@@ -108,6 +110,11 @@ def make_schema(endpoints: Tuple[str, ...]) -> Dict:
             }
         elif endpoint == "flaky":
             schema = {"parameters": [{"name": "id", "in": "query", "required": True, "type": "integer"}]}
+        elif endpoint == "path_variable":
+            schema = {
+                "parameters": [{"name": "key", "in": "path", "required": True, "type": "string", "minLength": 1}],
+                "responses": {200: {"description": "OK"}},
+            }
         elif endpoint == "invalid":
             schema = {"parameters": [{"name": "id", "in": "query", "required": True, "type": "int"}]}
         elif endpoint == "custom_format":
@@ -137,7 +144,7 @@ def make_schema(endpoints: Tuple[str, ...]) -> Dict:
                     "default": {"description": "Default response"},
                 }
             }
-        template["paths"][f"/{endpoint}"] = {method: schema}
+        template["paths"][path] = {method.lower(): schema}
     return template
 
 

--- a/test/test_runner.py
+++ b/test/test_runner.py
@@ -242,6 +242,16 @@ def test_response_conformance_malformed_json(schema_url, app):
     assert "Expecting property name enclosed in double quotes" in str(error)
 
 
+@pytest.mark.endpoints("path_variable")
+def test_path_parameters_encoding(schema_url, app):
+    # When endpoint has a path parameter
+    results = execute(schema_url, checks=(status_code_conformance,), hypothesis_options={"derandomize": True})
+    # Then there should be no failures
+    # since all path parameters are quoted
+    assert not results.has_errors
+    assert not results.has_failures
+
+
 @pytest.mark.parametrize(
     "options", ({"loader_options": {"base_url": "http://127.0.0.1:1/"}}, {"hypothesis_options": {"deadline": 1}})
 )

